### PR TITLE
Replace "gcontact" with "contact" - improved suggestions

### PR DESCRIPTION
**Pull request by annando**
In this release we will replace the "gcontact" table with the "contact" table. This is one of the steps towards this goal. The suggestions now work based on the "contact" table. Also the suggestion mechanism has completely changed. It should now work in most situations mostly reliable.

### CODE REVIEW

File: src/Model/Contact.php [hunk 1]
```diff
@@ -2915,4 +2915,157 @@ public static function canReceivePrivateMessages(array $contact)
...
+		$cid = self::getPublicIdByUserId($uid);
+		$totallimit = $start + $limit;
+		$contacts = [];
+
+		Logger::info('Collecting suggestions', ['uid' => $uid, 'cid' => $cid, 'start' => $start, 'limit' => $limit]);
+
+		$diaspora = DI::config()->get('system', 'diaspora_enabled') ? Protocol::DIASPORA : Protocol::ACTIVITYPUB;
+		$ostatus = !DI::config()->get('system', 'ostatus_disabled') ? Protocol::OSTATUS : Protocol::ACTIVITYPUB;
+
+		// The query returns contacts where contacts interacted with who the given user follows.
```
Comment (MrPetovan) [inline src/Model/Contact.php:2985]:
```suggestion
		// The query returns contacts where contacts interacted with whom the given user follows.
```

File: src/Model/Contact.php [hunk 2]
```diff
@@ -2915,4 +2915,157 @@ public static function canReceivePrivateMessages(array $contact)
...
+		}
+		DBA::close($results);
+
+		Logger::info('Contacts of contacts who are followed by the given user', ['uid' => $uid, 'cid' => $cid, 'count' => count($contacts)]);
+
+		if (count($contacts) >= $totallimit) {
+			return array_slice($contacts, $start, $limit);
+		}
+
+		// The query returns contacts where contacts interacted with who also interacted with the given user.
```
Comment (MrPetovan) [inline src/Model/Contact.php:3009]:
```suggestion
		// The query returns contacts where contacts interacted with whom also interacted with the given user.
```

**Comment by annando** [comment]
@MrPetovan okay, changes accepted.

File: src/Model/Contact.php [hunk 3]
```diff
@@ -2915,4 +2915,157 @@ public static function canReceivePrivateMessages(array $contact)
...
+		$ostatus = !DI::config()->get('system', 'ostatus_disabled') ? Protocol::OSTATUS : Protocol::ACTIVITYPUB;
+
+		// The query returns contacts where contacts interacted with whom the given user follows.
+		// Contacts who already are in the user's contact table are ignored.
+		$results = DBA::select('contact', [],
+			["`id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` IN
+				(SELECT `cid` FROM `contact-relation` WHERE `relation-cid` = ?)
+				AND NOT `cid` IN (SELECT `id` FROM `contact` WHERE `uid` = ? AND `nurl` IN
+					(SELECT `nurl` FROM `contact` WHERE `uid` = ? AND `rel` IN (?, ?))))
+			AND NOT `hidden` AND `network` IN (?, ?, ?, ?)",
```
Comment (MrPetovan) [inline src/Model/Contact.php:2992]: This SQL condition is absolutely impossible to parse as is. Can you please increase indentation each time you open a new sub-query?

File: src/Model/Contact.php [hunk 4]
```diff
@@ -2915,4 +2915,157 @@ public static function canReceivePrivateMessages(array $contact)
...
+		}
+
+		// The query returns contacts where contacts interacted with whom also interacted with the given user.
+		// Contacts who already are in the user's contact table are ignored.
+		$results = DBA::select('contact', [],
+			["`id` IN (SELECT `cid` FROM `contact-relation` WHERE `relation-cid` IN
+				(SELECT `relation-cid` FROM `contact-relation` WHERE `cid` = ?)
+				AND NOT `cid` IN (SELECT `id` FROM `contact` WHERE `uid` = ? AND `nurl` IN
+					(SELECT `nurl` FROM `contact` WHERE `uid` = ? AND `rel` IN (?, ?))))
+			AND NOT `hidden` AND `network` IN (?, ?, ?, ?)",
```
Comment (MrPetovan) [inline src/Model/Contact.php:3016]: Same ask for this query.

**Pull request merged**